src/search_controller.ts

```diff
@@ -0,0 +1,484 @@
+  isActive: boolean;
+  isLoading: boolean;
+  items: T[] | undefined;
+  searchQuery: string;
+  lastQueryError?: Error;
+  next?: string;
+  offset?: number;
+};
+
+export type SearchSourceOptions = {
+  /** The number of milliseconds to debounce the search query. The default interval is 300ms. */
+  debounceMs?: number;
+  isActive?: boolean;
+  pageSize?: number;
+};
+
+const DEFAULT_SEARCH_SOURCE_OPTIONS: Required<SearchSourceOptions> = {
+  debounceMs: 300,
+  isActive: false,
+  pageSize: 10,
+} as const;
+
+export abstract class BaseSearchSource<T> implements SearchSource<T> {
+  state: StateStore<SearchSourceState<T>>;
+  protected pageSize: number;
+  abstract readonly type: SearchSourceType;
+  searchDebounced!: DebouncedExecQueryFunction;
+
+  protected constructor(options?: SearchSourceOptions) {
+    const { debounceMs, isActive, pageSize } = { ...DEFAULT_SEARCH_SOURCE_OPTIONS, ...options };
+    this.pageSize = pageSize;
+    this.state = new StateStore<SearchSourceState<T>>({
+      hasMore: true,
+      isActive,
+      isLoading: false,
+      items: undefined,
+      offset: 0,
+      searchQuery: '',
+    });
+    this.setDebounceOptions({ debounceMs });
+  }
+
+  get lastQueryError() {
+    return this.state.getLatestValue().lastQueryError;
+  }
+
+  get hasMore() {
+    return this.state.getLatestValue().hasMore;
+  }
+
+  get hasResults() {
+    return Array.isArray(this.state.getLatestValue().items);
+  }
+
+  get isActive() {
+    return this.state.getLatestValue().isActive;
+  }
+
+  get isLoading() {
+    return this.state.getLatestValue().isLoading;
+  }
+
+  get items() {
+    return this.state.getLatestValue().items;
+  }
+
+  get next() {
+    return this.state.getLatestValue().next;
+  }
+
+  get offset() {
+    return this.state.getLatestValue().offset;
+  }
+
+  get searchQuery() {
+    return this.state.getLatestValue().searchQuery;
+  }
+
+  protected abstract query(searchQuery: string): Promise<QueryReturnValue<T>>;
+
+  protected abstract filterQueryResults(items: T[]): T[] | Promise<T[]>;
+
+  setDebounceOptions = ({ debounceMs }: DebounceOptions) => {
+    this.searchDebounced = debounce(this.executeQuery.bind(this), debounceMs);
+  };
+
+  activate = (sourceStateOverride?: Partial<SearchSourceState>) => {
+    if (this.isActive) return;
+    if (this.searchQuery) {
+      this.search();
+    }
+    this.state.partialNext({ ...sourceStateOverride, isActive: true });
+  };
+
+  deactivate = () => {
+    if (!this.isActive) return;
+    this.state.partialNext({ isActive: false });
+  };
+
+  async executeQuery(newSearchString?: string) {
+    const hasNewSearchQuery = typeof newSearchString !== 'undefined';
+    const searchString = newSearchString ?? this.searchQuery;
+    if (!this.isActive || this.isLoading || (!this.hasMore && !hasNewSearchQuery) || !searchString) return;
+
+    if (hasNewSearchQuery) {
+      this.resetState({ isActive: this.isActive, isLoading: true, searchQuery: newSearchString });
```

Comment:
To keep the behavior similar across the newly released classes, `resetState` shouldn't take any value but rather set the store to its initial state (at class construction). Any other modification should happen after the fact:

```suggestion
    if (hasNewSearchQuery) {
      this.resetState();
      this.state.partialNext({ isLoading: true, searchQuery: newSearchString })
```

Reply:
Or if you want to avoid two calls:

```ts
if (hasNewSearchQuery) {
    this.state.next((current) => {
        ...current,
        ...this.initialState,
        // overrides
        isActive: current.isActive,
        isLoading: true,
        searchQuery: newSearchString
    });
```